cocotb/triggers.py

```diff
@@ -544,3 +463,130 @@ def prime(self, callback):
 
     def __str__(self):
         return self.__class__.__name__ + "(%s)" % self._coroutine.__name__
+
+
+class Waitable(object):
+    """
+    Compatibility layer that emulates `collections.abc.Awaitable`.
+
+    This converts a `_wait` abstract method into a suitable `__await__` on
+    supporting python versions (>=3.3).
+    """
+    @decorators.coroutine
+    def _wait(self):
+        """
+        Should be implemented by the subclass. Called by `yield self` to
+        convert the waitable object into a coroutine.
+
+        ReturnValue can be used here
+        """
+        raise NotImplementedError
+        yield
+
+    if sys.version_info >= (3, 3):
+        def __await__(self):
+            return self._wait().__await__()
+
+
+class _AggregateWaitable(Waitable):
+    """
+    Base class for Waitables that take mutiple triggers in their constructor
+    """
+    def __init__(self, *args):
+        self.triggers = tuple(args)
+
+        # Do some basic type-checking up front, rather than waiting until we
+        # yield them.z
+        allowed_types = (Trigger, Waitable, decorators.RunningCoroutine)
+        for trigger in self.triggers:
+            if not isinstance(trigger, allowed_types):
+                raise TypeError(
+                    "All triggers must be instances of Trigger! Got: {}"
+                    .format(type(trigger).__name__)
+                )
+
+
+class Combine(_AggregateWaitable):
+    """
+    Combines multiple triggers together.
+
+    Waits until all triggers have fired.
+    """
+    @decorators.coroutine
+    def _wait(self):
+        waiters = []
+        e = Event()
+        triggers = list(self.triggers)
+
+        # start a parallel task for each trigger
+        for t in triggers:
+            @cocotb.coroutine
+            def waiter(t=t):
+                try:
+                    yield t
+                finally:
+                    triggers.remove(t)
+                    if not triggers:
+                        e.set()
+            waiters.append(cocotb.fork(waiter()))
+
+        # wait for the last waiter to complete
+        yield e.wait()
+        raise ReturnValue(self)
+
+
+class First(_AggregateWaitable):
+    """
+    Combines multiple triggers together.
+
+    Waits until the first trigger fires, and returns its result.
+    """
```

Comment:
Please also document what happens to triggers that didn't fire. And what is the behavior if two triggers fire at the same time?

Reply:
I don't think "at the same time" is a well-defined concept in an event loop - at best there's a concept of "within the same simulation tick", but the events still fire in a defined order.

Reply:
So what is the defined order from a user perspective? And please add a note that the non-first waitables are being killed.

Reply:
I have another small stylistic comment - in https://www.python.org/dev/peps/pep-0257/ they say 
> [The docstring is a phrase ending in a period.] It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ...".

Reply:
> And please add a note that the non-first waitables are being killed.



They're not - killing isn't transitive, the only thing that is killed is the internal coroutine used to wait on the things passed in. Killing a coroutine does not kill any nested calls currently in progress.

Reply:
@cmarque: I'd argue that this is not a function docstring, so it doesn't make sense to document as a command.